In [4]:
import pickle
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import confusion_matrix, roc_curve, roc_auc_score
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestClassifier

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
#importing the data
X,y = pickle.load(open("/content/drive/My Drive/ML_Fundamentals/cosmic.p","rb"))

#Q6.1

In [6]:
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 1, 0],
       ...,
       [0, 1, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int8)

In [7]:
y

array([4, 3, 1, ..., 2, 3, 1], dtype=int8)

In [8]:
#data exploration
X.shape

(6755, 18703)

In [9]:
y.shape

(6755,)

#Q6.1a

In [10]:
#How many of the 6,755 samples belong to each of the six types of cancer?
print("number of type 1 cancer samples",(y == 1).sum())
print("number of type 1 cancer samples",(y == 2).sum())
print("number of type 1 cancer samples",(y == 3).sum())
print("number of type 1 cancer samples",(y == 4).sum())
print("number of type 1 cancer samples",(y == 5).sum())
print("number of type 1 cancer samples",(y == 6).sum())

number of type 1 cancer samples 608
number of type 1 cancer samples 1867
number of type 1 cancer samples 1399
number of type 1 cancer samples 1363
number of type 1 cancer samples 1006
number of type 1 cancer samples 512


#Q6.1b

In [11]:
#count the total number of mutations for each sample (i.e. each row of X). What’s the smallest number? The largest number? The average?
max=0
min=18705
sum=0
t=0
for i in range(0, 6755):
  t=(X[i]==1).sum()
  if(t>max):
    max=t
  if(t<min):
    min=t
  sum=sum+t
  t=0



In [12]:
print("minimum mutations for each sample",min)

minimum mutations for each sample 45


In [13]:
print("maximum mutations for each sample",max)

maximum mutations for each sample 5938


In [14]:
print("average mutations for each sample",sum/6755)

average mutations for each sample 241.75292376017765


#Q6.1c

In [15]:
#count the total number of mutations for each gene (i.e. each column of X). What’s the smallest number? The largest number? The average?
max=0
min=18705
sum=0
t=0

for i in range(0,18703):
  for j in range(0,6755):
    t=t+X[j][i]
  if(t>max):
    max=t
  if(t<min):
    min=t
  sum=sum+t
  t=0

In [16]:
print("minimum mutations for each gene",min)

minimum mutations for each gene 30


In [17]:
print("maximum mutations for each gene",max)

maximum mutations for each gene 1768


In [18]:
print("average mutations for each gene",sum/18703)

average mutations for each gene 87.31438806608566


In [104]:
#Treat the first 4,500 rows of the data as your training set (and the remaining 2,255 rows as your test set)
X_train = X[:4500]

In [105]:
X_train.shape

(4500, 18703)

In [106]:
X_test = X[4501:]

In [107]:
X_test.shape

(2254, 18703)

In [108]:
y_train = y[:4500]

In [109]:
y_train.shape

(4500,)

In [110]:
y_test=y[4501:]

In [111]:
y_test.shape

(2254,)

#Q6.2a

##logistic regression

In [112]:
#logistic regression binary classifier for cancer type 1
binarylabels1 = np.where(y_train == 1, 1, 0)
mdl1 = LogisticRegression()
mdl1.fit(X_train, binarylabels1)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl1.predict_proba(X_test)
threshold = 0.5
y_pred = mdl1.predict_proba(X_test)[:,1]
y_test1 = np.where(y_test == 1, 1, 0)
roc_auc_score(y_test1,y_pred)

0.9607288579164498

In [113]:
#logistic regression binary classifier for cancer type 2
binarylabels2 = np.where(y_train == 2, 1, 0)
mdl2 = LogisticRegression()
mdl2.fit(X_train, binarylabels2)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl2.predict_proba(X_test)
threshold = 0.5
y_pred = mdl2.predict_proba(X_test)[:,1]
y_test2 = np.where(y_test == 2, 1, 0)
roc_auc_score(y_test2,y_pred)

0.9576842362724716

In [114]:
#logistic regression binary classifier for cancer type 3
binarylabels3 = np.where(y_train == 3, 1, 0)
mdl3 = LogisticRegression()
mdl3.fit(X_train, binarylabels3)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl3.predict_proba(X_test)
threshold = 0.5
y_pred = mdl3.predict_proba(X_test)[:,1]
y_test3 = np.where(y_test == 3, 1, 0)
roc_auc_score(y_test3,y_pred)

0.9252122620865386

In [115]:
#logistic regression binary classifier for cancer type 4
binarylabels4 = np.where(y_train == 4, 1, 0)
mdl4 = LogisticRegression()
mdl4.fit(X_train, binarylabels4)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl4.predict_proba(X_test)
threshold = 0.5
y_pred = mdl4.predict_proba(X_test)[:,1]
y_test4 = np.where(y_test == 4, 1, 0)
roc_auc_score(y_test4,y_pred)

0.9367978719871906

In [116]:
#logistic regression binary classifier for cancer type 5
binarylabels5 = np.where(y_train == 5, 1, 0)
mdl5 = LogisticRegression()
mdl5.fit(X_train, binarylabels5)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl5.predict_proba(X_test)
threshold = 0.5
y_pred = mdl5.predict_proba(X_test)[:,1]
y_test5 = np.where(y_test == 5, 1, 0)
roc_auc_score(y_test5,y_pred)

0.8852794508326269

In [117]:
#logistic regression binary classifier for cancer type 6
binarylabels6 = np.where(y_train == 6, 1, 0)
mdl6 = LogisticRegression()
mdl6.fit(X_train, binarylabels6)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl6.predict_proba(X_test)
threshold = 0.5
y_pred = mdl6.predict_proba(X_test)[:,1]
y_test6 = np.where(y_test == 6, 1, 0)
roc_auc_score(y_test6,y_pred)

0.9151109508252365

#Q6.2b

##logistic regression with L1 penalty

In [118]:
#logistic regression binary classifier for cancer type 1
binarylabels1 = np.where(y_train == 1, 1, 0)
mdl1 = LogisticRegression(penalty='l1', solver='liblinear')
mdl1.fit(X_train, binarylabels1)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl1.predict_proba(X_test)
threshold = 0.5
y_pred = mdl1.predict_proba(X_test)[:,1]
y_test1 = np.where(y_test == 1, 1, 0)
roc_auc_score(y_test1,y_pred)

0.9484266151800148

In [119]:
#logistic regression binary classifier for cancer type 2
binarylabels2 = np.where(y_train == 2, 1, 0)
mdl2 = LogisticRegression(penalty='l1', solver='liblinear')
mdl2.fit(X_train, binarylabels2)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl2.predict_proba(X_test)
threshold = 0.5
y_pred = mdl2.predict_proba(X_test)[:,1]
y_test2 = np.where(y_test == 2, 1, 0)
roc_auc_score(y_test2,y_pred)

0.9437231258407727

In [120]:
#logistic regression binary classifier for cancer type 3
binarylabels3 = np.where(y_train == 3, 1, 0)
mdl3 = LogisticRegression(penalty='l1', solver='liblinear')
mdl3.fit(X_train, binarylabels3)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl3.predict_proba(X_test)
threshold = 0.5
y_pred = mdl3.predict_proba(X_test)[:,1]
y_test3 = np.where(y_test == 3, 1, 0)
roc_auc_score(y_test3,y_pred)

0.8943491601910115

In [121]:
#logistic regression binary classifier for cancer type 4
binarylabels4 = np.where(y_train == 4, 1, 0)
mdl4 = LogisticRegression(penalty='l1', solver='liblinear')
mdl4.fit(X_train, binarylabels4)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl4.predict_proba(X_test)
threshold = 0.5
y_pred = mdl4.predict_proba(X_test)[:,1]
y_test4 = np.where(y_test == 4, 1, 0)
roc_auc_score(y_test4,y_pred)

0.906780150844921

In [122]:
#logistic regression binary classifier for cancer type 5
binarylabels5 = np.where(y_train == 5, 1, 0)
mdl5 = LogisticRegression(penalty='l1', solver='liblinear')
mdl5.fit(X_train, binarylabels5)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl5.predict_proba(X_test)
threshold = 0.5
y_pred = mdl5.predict_proba(X_test)[:,1]
y_test5 = np.where(y_test == 5, 1, 0)
roc_auc_score(y_test5,y_pred)

0.8858573765239532

In [123]:
#logistic regression binary classifier for cancer type 6
binarylabels6 = np.where(y_train == 6, 1, 0)
mdl6 = LogisticRegression(penalty='l1', solver='liblinear')
mdl6.fit(X_train, binarylabels6)
# clf = (penalty='none',max_iter=1000LogisticRegression)
mdl6.predict_proba(X_test)
threshold = 0.5
y_pred = mdl6.predict_proba(X_test)[:,1]
y_test6 = np.where(y_test == 6, 1, 0)
roc_auc_score(y_test6,y_pred)

0.9080439348296492

#Q6.2c

##random_forest

In [124]:
mdl1 = RandomForestClassifier(random_state=10)
mdl1.fit(X_train,binarylabels1)

y_pred = mdl1.predict_proba(X_test)[:,1]
roc_auc_score(y_test1,y_pred)

0.8220065213073788

In [125]:
mdl2 = RandomForestClassifier(random_state=10)
mdl2.fit(X_train,binarylabels2)

y_pred = mdl2.predict_proba(X_test)[:,1]
roc_auc_score(y_test2,y_pred)

0.9228783172312585

In [126]:
mdl3 = RandomForestClassifier(random_state=10)
mdl3.fit(X_train,binarylabels3)

y_pred = mdl3.predict_proba(X_test)[:,1]
roc_auc_score(y_test3,y_pred)

0.8305465323231636

In [127]:
mdl4 = RandomForestClassifier(random_state=10)
mdl4.fit(X_train,binarylabels4)

y_pred = mdl4.predict_proba(X_test)[:,1]
roc_auc_score(y_test4,y_pred)

0.8761241734335359

In [128]:
mdl5 = RandomForestClassifier(random_state=10)
mdl5.fit(X_train,binarylabels5)

y_pred = mdl5.predict_proba(X_test)[:,1]
roc_auc_score(y_test5,y_pred)

0.7578882257987868

In [129]:
mdl6 = RandomForestClassifier(random_state=10)
mdl6.fit(X_train,binarylabels6)

y_pred = mdl6.predict_proba(X_test)[:,1]
roc_auc_score(y_test6,y_pred)

0.8357402732402732

#PCA

In [52]:
#setting the predictor variable to 1 where cancer type=1
# y1 = np.where(y == 1, 1, 0)
# max_auc=0
# best_no_of_comp = 0
# auc=0
# comp_values = [2500,2000,1500,1000,500] #the number of components I want to try
# for n_components in comp_values: #looping over the 4 number of components
#   pca = PCA(n_components=n_components) #defining the PCA function
#   X_pca = pca.fit_transform(X) #pca on X
#   X_pca_train = X_pca[:3000] # X split to train
#   X_pca_val = X_pca[3000:4500] #X split to val
#   X_pca_test = X_pca[4500:] #X split to test
#   y_train = y1[:3000] #y1 split to train
#   y_val = y1[3000:4500] #y1 split to val
#   y_test = y1[4500:] #y1 split to test
#   #logistic regression binary classifier for cancer type 1
#   mdl1pca = LogisticRegression() #initiating the logistic regression
#   mdl1pca.fit(X_pca_train, y_train) #fit model
#   mdl1pca.predict_proba(X_pca_val)
#   threshold = 0.5
#   y_pred_val = mdl1pca.predict_proba(X_pca_val)[:,1]
#   ans = roc_auc_score(y_val,y_pred_val)
#   if ans>max_auc:
#     max_auc=ans
#     best_no_of_comp=n_components
#     y_pred_test=mdl1pca.predict_proba(X_pca_test)[:,1]
#     auc = roc_auc_score(y_test,y_pred_test)
#   print("auc_score for # components = ", n_components)
#   print(ans)

# print("AUC score on the test set(out_of_sample_auc) with ",best_no_of_comp, " is ",auc)

auc_score for # components =  2500
0.938502203675919
auc_score for # components =  2000
0.9316508814703675
auc_score for # components =  1500
0.9334325768942235
auc_score for # components =  1000
0.9380685015003752
auc_score for # components =  500
0.9146075581395349
AUC score on the test set(out_of_sample_auc) with  2500  is  0.947163828065714


In [130]:
# comp_values = [3000,2500,2000] #the number of components I want to try
# for n_components in comp_values: #looping over the 4 number of components
#   pca3000 = PCA(n_components=3000) #defining the PCA function
#   X_pca_train = X[:3000] # X split to train
#   X_pca_val = X[3000:4500] #X split to val
#   X_pca_test = X[4500:] #X split to test
#   X_pca_train = pca.fit_transform(X) #pca on X
#   print("auc's for comp=",n_components)
  # X_pca_train_3000 =
X_pca_train = X[:3000] # X split to train
X_pca_val = X[3000:4500] #X split to val
X_pca_test = X[4500:] #X split to test
pca3000 = PCA(n_components=3000)
X_pca_train_3000 = pca3000.fit_transform(X_pca_train)
X_pca_val_3000 = pca3000.transform(X_pca_val)
# X_pca_test_3000 = pca3000.transform(X_pca_test)
pca2500 = PCA(n_components=2500)
X_pca_train_2500 = pca2500.fit_transform(X_pca_train)
X_pca_val_2500 = pca2500.transform(X_pca_val)
# X_pca_test_2500 = pca2500.transform(X_pca_test)
pca2000 = PCA(n_components=2000)
X_pca_train_2000 = pca2000.fit_transform(X_pca_train)
X_pca_val_2000 = pca2000.transform(X_pca_val)
# X_pca_test_2000 = pca2000.transform(X_pca_test)




#Q6.3a

##logistic regression with 2000, 2500 and 3000 Principle Components

In [ ]:
for i in range(1,7):
  yi = np.where(y == i, 1, 0)
  max_auc=0

  y_train = yi[:3000] #y1 split to train
  y_val = yi[3000:4500] #y1 split to val
  y_test = yi[4500:] #y1 split to test
  mdlpca3000 = LogisticRegression() #initiating the logistic regression
  mdlpca3000.fit(X_pca_train_3000, y_train) #fit model
  mdlpca3000.predict_proba(X_pca_val_3000)
  threshold = 0.5
  y_pred_val3000 = mdlpca3000.predict_proba(X_pca_val_3000)[:,1]
  ans3000=roc_auc_score(y_val,y_pred_val3000)
  print("auc_score for 3000 components = ", ans3000)
  if ans3000>max_auc:
    max_auc=ans3000
    best_no_of_comp=3000
    mdl=mdlpca3000
    pca=pca3000

  mdlpca2500 = LogisticRegression() #initiating the logistic regression
  mdlpca2500.fit(X_pca_train_2500, y_train) #fit model
  mdlpca2500.predict_proba(X_pca_val_2500)
  threshold = 0.5
  y_pred_val2500 = mdlpca2500.predict_proba(X_pca_val_2500)[:,1]
  ans2500=roc_auc_score(y_val,y_pred_val2500)
  print("auc_score for 2500 components = ", ans2500)
  if ans2500>max_auc:
    max_auc=ans2500
    best_no_of_comp=2500
    mdl=mdlpca2500
    pca=pca2500

  mdlpca2000 = LogisticRegression() #initiating the logistic regression
  mdlpca2000.fit(X_pca_train_2000, y_train) #fit model
  mdlpca2000.predict_proba(X_pca_val_2000)
  threshold = 0.5
  y_pred_val2000 = mdlpca2000.predict_proba(X_pca_val_2000)[:,1]
  ans2000=roc_auc_score(y_val,y_pred_val2000)
  print("auc_score for 2000 components = ", ans2000)
  if ans2000>max_auc:
    max_auc=ans2000
    best_no_of_comp=2000
    mdl=mdlpca2000
    pca=pca2000

  X_pca_test_components = pca.transform(X_pca_test)
  y_pred_test=mdl.predict_proba(X_pca_test_components)[:,1]
  print("Out of Sample AUC score is ", roc_auc_score(y_test,y_pred_test), " for number of components= ", best_no_of_comp)

#Q6.3b

##logistic regression with L1 penalty and 2000,3000,2500 principle components

In [59]:
for i in range(1,7):
  yi = np.where(y == i, 1, 0)
  max_auc=0

  y_train = yi[:3000] #y1 split to train
  y_val = yi[3000:4500] #y1 split to val
  y_test = yi[4500:] #y1 split to test
  mdlpca3000 = LogisticRegression(penalty='l1', solver='liblinear') #initiating the logistic regression
  mdlpca3000.fit(X_pca_train_3000, y_train) #fit model
  mdlpca3000.predict_proba(X_pca_val_3000)
  threshold = 0.5
  y_pred_val3000 = mdlpca3000.predict_proba(X_pca_val_3000)[:,1]
  ans3000=roc_auc_score(y_val,y_pred_val3000)
  print("auc_score for 3000 components = ", ans3000)
  if ans3000>max_auc:
    max_auc=ans3000
    best_no_of_comp=3000
    mdl=mdlpca3000
    pca=pca3000

  mdlpca2500 = LogisticRegression(penalty='l1', solver='liblinear') #initiating the logistic regression
  mdlpca2500.fit(X_pca_train_2500, y_train) #fit model
  mdlpca2500.predict_proba(X_pca_val_2500)
  threshold = 0.5
  y_pred_val2500 = mdlpca2500.predict_proba(X_pca_val_2500)[:,1]
  ans2500=roc_auc_score(y_val,y_pred_val2500)
  print("auc_score for 2500 components = ", ans2500)
  if ans2500>max_auc:
    max_auc=ans2500
    best_no_of_comp=2500
    mdl=mdlpca2500
    pca=pca2500

  mdlpca2000 = LogisticRegression(penalty='l1', solver='liblinear') #initiating the logistic regression
  mdlpca2000.fit(X_pca_train_2000, y_train) #fit model
  mdlpca2000.predict_proba(X_pca_val_2000)
  threshold = 0.5
  y_pred_val2000 = mdlpca2000.predict_proba(X_pca_val_2000)[:,1]
  ans2000=roc_auc_score(y_val,y_pred_val2000)
  print("auc_score for 2000 components = ", ans2000)
  if ans2000>max_auc:
    max_auc=ans2000
    best_no_of_comp=2000
    mdl=mdlpca2000
    pca=pca2000

  X_pca_test_components = pca.transform(X_pca_test)
  y_pred_test=mdl.predict_proba(X_pca_test_components)[:,1]
  print("Out of Sample AUC score is ", roc_auc_score(y_test,y_pred_test), " for number of components= ", best_no_of_comp)

auc_score for 3000 components =  0.9206911102775693
auc_score for 2500 components =  0.9200933045761441
auc_score for 2000 components =  0.9166764347336833
Out of Sample AUC score is  0.9374418920209264  for number of components=  3000
auc_score for 3000 components =  0.9543276807665637
auc_score for 2500 components =  0.9531411820858962
auc_score for 2000 components =  0.9546338739744779
Out of Sample AUC score is  0.950726176578141  for number of components=  2000
auc_score for 3000 components =  0.876182801078482
auc_score for 2500 components =  0.8793793049361915
auc_score for 2000 components =  0.8777026277857857
Out of Sample AUC score is  0.87941491865244  for number of components=  2500
auc_score for 3000 components =  0.9152124881291548
auc_score for 2500 components =  0.9116134636967971
auc_score for 2000 components =  0.9064739273072606
Out of Sample AUC score is  0.9204145997315648  for number of components=  3000
auc_score for 3000 components =  0.8477209353422233
auc_scor

#Q6.3c

##random forest with 2000,3000 and 2500 principle components

In [131]:
for i in range(1,7):
  yi = np.where(y == i, 1, 0)
  max_auc=0

  y_train = yi[:3000] #y1 split to train
  y_val = yi[3000:4500] #y1 split to val
  y_test = yi[4500:] #y1 split to test
  mdlpca3000 = RandomForestClassifier(random_state=10) #initiating the logistic regression
  mdlpca3000.fit(X_pca_train_3000, y_train) #fit model
  mdlpca3000.predict_proba(X_pca_val_3000)
  threshold = 0.5
  y_pred_val3000 = mdlpca3000.predict_proba(X_pca_val_3000)[:,1]
  ans3000=roc_auc_score(y_val,y_pred_val3000)
  print("auc_score for 3000 components = ", ans3000)
  if ans3000>max_auc:
    max_auc=ans3000
    best_no_of_comp=3000
    mdl=mdlpca3000
    pca=pca3000

  mdlpca2500 = RandomForestClassifier(random_state=10) #initiating the logistic regression
  mdlpca2500.fit(X_pca_train_2500, y_train) #fit model
  mdlpca2500.predict_proba(X_pca_val_2500)
  threshold = 0.5
  y_pred_val2500 = mdlpca2500.predict_proba(X_pca_val_2500)[:,1]
  ans2500=roc_auc_score(y_val,y_pred_val2500)
  print("auc_score for 2500 components = ", ans2500)
  if ans2500>max_auc:
    max_auc=ans2500
    best_no_of_comp=2500
    mdl=mdlpca2500
    pca=pca2500

  mdlpca2000 = RandomForestClassifier(random_state=10) #initiating the logistic regression
  mdlpca2000.fit(X_pca_train_2000, y_train) #fit model
  mdlpca2000.predict_proba(X_pca_val_2000)
  threshold = 0.5
  y_pred_val2000 = mdlpca2000.predict_proba(X_pca_val_2000)[:,1]
  ans2000=roc_auc_score(y_val,y_pred_val2000)
  print("auc_score for 2000 components = ", ans2000)
  if ans2000>max_auc:
    max_auc=ans2000
    best_no_of_comp=2000
    mdl=mdlpca2000
    pca=pca2000

  X_pca_test_components = pca.transform(X_pca_test)
  y_pred_test=mdl.predict_proba(X_pca_test_components)[:,1]
  print("Out of Sample AUC score is ", roc_auc_score(y_test,y_pred_test), " for number of components= ", best_no_of_comp)

auc_score for 3000 components =  0.3679875046886722
auc_score for 2500 components =  0.4907574549887472
auc_score for 2000 components =  0.8023929810577645
Out of Sample AUC score is  0.8006349788675414  for number of components=  2000
auc_score for 3000 components =  0.7612762403076341
auc_score for 2500 components =  0.7850230995758323
auc_score for 2000 components =  0.7781686494177826
Out of Sample AUC score is  0.7780595333775276  for number of components=  2500
auc_score for 3000 components =  0.563410848101163
auc_score for 2500 components =  0.6376565462073293
auc_score for 2000 components =  0.6734549555274907
Out of Sample AUC score is  0.6786668267901073  for number of components=  2000
auc_score for 3000 components =  0.8293161313994648
auc_score for 2500 components =  0.8341076361909695
auc_score for 2000 components =  0.8410021691271692
Out of Sample AUC score is  0.844131788929258  for number of components=  2000
auc_score for 3000 components =  0.6119224791674072
auc_sc



####Now let's list all the out-of-sample AUC scores for each of the 6 cancer types for all 3 models with and without PCA

#Summary

#Cancer Type 1:


Logistic Regression:0.9607288579164498

Logistic Regression with L1 penalty: 0.9484266151800148

Random Forest: 0.8220065213073788


Logistic Regression with PCA(# components= 3000): 0.9574561645969757

Logistic Regression with L1 penalty with PCA(# components= 3000): 0.9374418920209264

Random Forest with PCA(# components= 2000): 0.800634978867541

###For predicting cancer type 1, a simple logistic regression model gives the best out-of-sample AUC score, so we should use a logistic regression model to predict type 1 cancer.


#Cancer Type 2:


Logistic Regression: 0.9576842362724716

Logistic Regression with L1 penalty: 0.9437231258407727

Random Forest: 0.9228783172312585


Logistic Regression with PCA(# components= 2000): 0.9506127573199419

Logistic Regression with L1 penalty with PCA(# components= 2000): 0.950726176578141

Random Forest with PCA(# components= 2500): 0.7780595333775276

###For predicting cancer type 2, a simple logistic regression model gives the best out-of-sample AUC score. So, we should use a logistic regression model to predict type 2 cancer.


#Cancer Type 3:


Logistic Regression: 0.9252122620865386

Logistic Regression with L1 penalty: 0.8943491601910115

Random Forest: 0.8305465323231636

Logistic Regression with PCA(# components= 2500): 0.903650571834772

Logistic Regression with L1 penalty with PCA(# components= 2500): 0.87941491865244

Random Forest with PCA(# components= 2000): 0.678666826790107

###For predicting cancer type 3, a simple logistic regression model gives the best out-of-sample AUC score, so we should use a logistic regression model to predict type 3 cancer.

#Cancer Type 4:

Logistic Regression: 0.9367978719871906

Logistic Regression with L1 penalty: 0.906780150844921

Random Forest: 0.8761241734335359

Logistic Regression with PCA(# components= 2000): 0.922406969483926

Logistic Regression with L1 penalty with PCA(# components= 3000): 0.9204145997315648

Random Forest with PCA(# components= 2000): 0.844131788929258

###For predicting cancer type 4, a simple logistic regression model gives the best out-of-sample AUC score, so we should use a logistic regression model to predict type 4 cancer.

#Cancer Type 5:

Logistic Regression: 0.8852794508326269

Logistic Regression with L1 penalty: 0.8858573765239532

Random Forest: 0.7578882257987868

Logistic Regression with PCA(# components= 2500): 0.8634632297287773

Logistic Regression with L1 penalty with PCA(# components= 2000): 0.8433996145089278

Random Forest with PCA(# components= 3000): 0.6253106451306472

###For predicting cancer type 5, a simple logistic regression model with L1 penalty gives the best out-of-sample AUC score, so we should use a logistic regression model with L1 penalty to predict type 5 cancer.

#Cancer Type 6:

Logistic Regression: 0.9151109508252365

Logistic Regression with L1 penalty: 0.9080439348296492

Random Forest: 0.8357402732402732

Logistic Regression with PCA(# components= 2000): 0.9097713670796159

Logistic Regression with L1 penalty with PCA(# components= 2500): 0.899020901915258

Random Forest with PCA(# components= 2000): 0.8164535657299767

###For predicting cancer type 6, a simple logistic regression model gives the best out-of-sample AUC score, so we should use a logistic regression model to predict type 6 cancer.




#Q6.4

The first method could be to take the average of all the AUC scores for all 6 of the cancer types





In [20]:
(0.9607288579164498+0.9576842362724716+0.9252122620865386+0.9367978719871906+0.8858573765239532+0.9151109508252365)/6

0.9302319259353067

The average AUC over all 6 cancer types is 0.9302319259353067.

I feel another method to find the average AUC for the whole model together we should take the weighted average of highest AUC's for all of the 6 types of cancer, since the number of samples is different in each of the 6 cancer types.

Also, for 5 of the 6 cancer types, the simple logistic model performs the best, so we can just use a logistic regression model over all 6 cancer types to find the average AUC.